File: examples/vulnerable_code/test_vulnerabilities.js
/**
 * Example file with intentional security vulnerabilities for testing SAST scanning.
 *
 * DO NOT use this code in production!
 * This file is for testing and demonstration purposes only.
 */

const express = require('express');
const { exec } = require('child_process');
const crypto = require('crypto');
const app = express();

// ============================================================================
// VULNERABILITY 1: SQL Injection
// ============================================================================

function getUserVulnerable(userId) {
    // VULNERABLE: SQL injection via string concatenation
    // CWE-89: Improper Neutralization of Special Elements used in an SQL Command
    const query = `SELECT * FROM users WHERE id = ${userId}`; // VULNERABLE
    return db.execute(query);
}

function getUserSafe(userId) {
    // SAFE: Using parameterized query
    const query = 'SELECT * FROM users WHERE id = ?';
    return db.execute(query, [userId]);
}

// ============================================================================
// VULNERABILITY 2: Command Injection
// ============================================================================

function pingHostVulnerable(hostname) {
    // VULNERABLE: Command injection via exec()
    // CWE-77: Improper Neutralization of Special Elements used in an OS Command
    exec(`ping -c 4 ${hostname}`, (error, stdout) => { // VULNERABLE
        console.log(stdout);
    });
}

function pingHostSafe(hostname) {
    // SAFE: Using spawn with array arguments (no shell interpretation)
    const { spawn } = require('child_process');
    spawn('ping', ['ping', '-c', '4', hostname]);
}

// ============================================================================
// VULNERABILITY 3: Hardcoded Secrets
// ============================================================================

// VULNERABLE: Hardcoded API key
const API_KEY = 'test_api_key_replace_with_real_key_do_not_commit'; // VULNERABLE

// VULNERABLE: Hardcoded database password
const DB_PASSWORD = 'test_password_replace_with_real_password'; // VULNERABLE

function getApiConfigSafe() {
    // SAFE: Load from environment variables
    return {
        apiKey: process.env.API_KEY,
        dbPassword: process.env.DB_PASSWORD
    };
}

// ============================================================================
// VULNERABILITY 4: Cross-Site Scripting (XSS)
// ============================================================================

app.get('/greeting', (req, res) => {
    const username = req.query.username;
    // VULNERABLE: Reflected XSS via unsanitized user input
    // CWE-79: Improper Neutralization of Input During Web Page Generation
    res.send(`<h1>Hello, ${username}!</h1>`); // VULNERABLE
});

app.get('/greeting-safe', (req, res) => {
    const username = req.query.username;
    // SAFE: Use template with auto-escaping or sanitize input
    const sanitizeHtml = require('sanitize-html');
    const clean = sanitizeHtml(username);
    res.send(`<h1>Hello, ${clean}!</h1>`);
});

// ============================================================================
// VULNERABILITY 5: eval() Usage - Code Injection
// ============================================================================

function calculateVulnerable(expression) {
    // VULNERABLE: Using eval() with user input
    // CWE-95: Improper Neutralization of Directives in Dynamically Evaluated Code
    return eval(expression); // VULNERABLE
}

function calculateSafe(expression) {
    // SAFE: Use a proper expression parser
    const { Parser } = require('expr-eval');
    const parser = new Parser();
    return parser.parse(expression).value;
}

// ============================================================================
// VULNERABILITY 6: Weak Cryptography
// ============================================================================

function hashPasswordVulnerable(password) {
    // VULNERABLE: Using MD5 for password hashing
    // CWE-327: Use of a Broken or Risky Cryptographic Algorithm
    return crypto.createHash('md5').update(password).digest('hex'); // VULNERABLE
}

function hashPasswordSafe(password) {
    // SAFE: Using bcrypt for password hashing
    const bcrypt = require('bcrypt');
    const salt = bcrypt.genSaltSync(10);
    return bcrypt.hashSync(password, salt);
}

// ============================================================================
// VULNERABILITY 7: Insecure Random Number Generation
// ============================================================================

function generateTokenVulnerable() {
    // VULNERABLE: Using Math.random() for security-sensitive tokens
    // CWE-338: Use of Cryptographically Weak PRNG
    let token = '';
    const chars = 'abcdef0123456789';
    for (let i = 0; i < 32; i++) {
        token += chars[Math.floor(Math.random() * chars.length)]; // VULNERABLE
    }
    return token;
}

function generateTokenSafe() {
    // SAFE: Using crypto.randomBytes() for cryptographic security
    return crypto.randomBytes(32).toString('hex');
}

// ============================================================================
// VULNERABILITY 8: Path Traversal
// ============================================================================

const fs = require('fs');

function readFileVulnerable(filename) {
    // VULNERABLE: Path traversal via unsanitized filename
    // CWE-22: Improper Limitation of a Pathname to a Restricted Directory
    const path = `/var/app/files/${filename}`; // VULNERABLE
    return fs.readFileSync(path, 'utf8');
}

function readFileSafe(filename) {
    // SAFE: Validate and sanitize filename
    const path = require('path');

    // Remove any path components
    const sanitized = filename.replace(/[^a-zA-Z0-9._-]/g, '');

    // Construct full path
    const fullPath = path.join('/var/app/files', sanitized);

    // Ensure result is within allowed directory
    const resolvedPath = path.resolve(fullPath);
    if (!resolvedPath.startsWith('/var/app/files/')) {
        throw new Error('Access denied');
    }

    return fs.readFileSync(resolvedPath, 'utf8');
}

// ============================================================================
// VULNERABILITY 9: Insecure Direct Object Reference (IDOR)
// ============================================================================

app.get('/orders/:orderId', (req, res) => {
    const orderId = req.params.orderId;
    const order = db.getOrder(orderId);
    // VULNERABLE: No authorization check - any user can access any order
    // CWE-639: Insecure Direct Object Reference
    res.json(order); // VULNERABLE
});

app.get('/orders-safe/:orderId', (req, res) => {
    const orderId = req.params.orderId;
    const order = db.getOrder(orderId);

    // SAFE: Check that user owns the order
    if (order.userId !== req.user.id) {
        return res.status(403).json({ error: 'Access denied' });
    }

    res.json(order);
});

// ============================================================================
// VULNERABILITY 10: Regular Expression Denial of Service (ReDoS)
// ============================================================================

function validateEmailVulnerable(email) {
    // VULNERABLE: Catastrophic backtracking regex
    // CWE-1333: Inefficient Regular Expression Complexity
    const regex = /^([a-zA-Z0-9])(([\-.]|[_]+)?([a-zA-Z0-9]+))*@(([a-zA-Z0-9]-)+(([a-zA-Z0-9])+))$/; // VULNERABLE
    return regex.test(email);
}

function validateEmailSafe(email) {
    // SAFE: Use a well-tested email validation library or simple regex
    const regex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return regex.test(email);
}

// ============================================================================
// VULNERABILITY 11: Sensitive Data in URL
// ============================================================================

app.get('/reset-password', (req, res) => {
    const token = req.query.token;
    // VULNERABLE: Sensitive token passed in URL (will be logged)
    // CWE-598: Use of GET Request Method With Sensitive Query Strings
    return res.send(`Reset password with token: ${token}`); // VULNERABLE
});

app.post('/reset-password', (req, res) => {
    const token = req.body.token; // SAFE: Use POST body for sensitive data
    // Validate token and reset password
    return res.send('Password reset email sent');
});

// ============================================================================
// VULNERABILITY 12: Missing Rate Limiting
// ============================================================================

app.post('/login', (req, res) => {
    const { username, password } = req.body;
    // VULNERABLE: No rate limiting - vulnerable to brute force attacks
    // CWE-307: Improper Restriction of Excessive Authentication Attempts
    const user = authenticateUser(username, password);
    if (user) {
        res.json({ token: user.token });
    } else {
        res.status(401).json({ error: 'Invalid credentials' });
    }
});

// SAFE: Use rate limiting middleware
const rateLimit = require('express-rate-limit');

const loginLimiter = rateLimit({
    windowMs: 15 * 60 * 1000, // 15 minutes
    max: 5, // limit each IP to 5 requests per windowMs
    message: 'Too many login attempts'
});

app.post('/login-safe', loginLimiter, (req, res) => {
    const { username, password } = req.body;
    const user = authenticateUser(username, password);
    if (user) {
        res.json({ token: user.token });
    } else {
        res.status(401).json({ error: 'Invalid credentials' });
    }
});

if (require.main === module) {
    console.log('This file contains intentional vulnerabilities for testing SAST scanning');
    console.log('Do NOT use in production!');
}

module.exports = app;

File: examples/vulnerable_code/test_vulnerabilities.py
"""
Example file with intentional security vulnerabilities for testing SAST scanning.

DO NOT use this code in production!
This file is for testing and demonstration purposes only.
"""

import os
import subprocess
import pickle
from flask import Flask, request, make_response

app = Flask(__name__)

# ============================================================================
# VULNERABILITY 1: SQL Injection
# ============================================================================

def get_user_vulnerable(user_id):
    """
    VULNERABLE: SQL Injection via string concatenation
    CWE-89: Improper Neutralization of Special Elements used in an SQL Command
    """
    query = f"SELECT * FROM users WHERE id = {user_id}"  # VULNERABLE
    return db.execute(query)


def get_user_safe(user_id):
    """
    SAFE: Using parameterized query
    """
    query = "SELECT * FROM users WHERE id = %s"
    return db.execute(query, (user_id,))


# ============================================================================
# VULNERABILITY 2: Command Injection
# ============================================================================

def ping_host_vulnerable(hostname):
    """
    VULNERABLE: Command injection via subprocess
    CWE-77: Improper Neutralization of Special Elements used in an OS Command
    """
    result = os.system(f"ping -c 4 {hostname}")  # VULNERABLE
    return result


def ping_host_safe(hostname):
    """
    SAFE: Using subprocess with list argument (no shell interpretation)
    """
    result = subprocess.run(['ping', '-c', '4', hostname], capture_output=True)
    return result


# ============================================================================
# VULNERABILITY 3: Hardcoded Secrets
# ============================================================================

# VULNERABLE: Hardcoded API key in source code
API_KEY = "test_api_key_replace_with_real_key_do_not_commit"  # VULNERABLE

# VULNERABLE: Hardcoded database credentials
DB_PASSWORD = "test_password_replace_with_real_password"  # VULNERABLE


def get_api_config_safe():
    """
    SAFE: Load secrets from environment variables
    """
    return {
        'api_key': os.environ.get('API_KEY'),
        'db_password': os.environ.get('DB_PASSWORD')
    }


# ============================================================================
# VULNERABILITY 4: Cross-Site Scripting (XSS)
# ============================================================================

@app.route('/greeting')
def greeting_vulnerable():
    """
    VULNERABLE: Reflected XSS via unsanitized user input
    CWE-79: Improper Neutralization of Input During Web Page Generation
    """
    username = request.args.get('username', '')
    html = f"<h1>Hello, {username}!</h1>"  # VULNERABLE
    return html


@app.route('/greeting-safe')
def greeting_safe():
    """
    SAFE: Using template with auto-escaping (Flask/Jinja2)
    """
    from flask import render_template_string
    username = request.args.get('username', '')
    return render_template_string("<h1>Hello, {{ username }}!</h1>", username=username)


# ============================================================================
# VULNERABILITY 5: Insecure Deserialization
# ============================================================================

@app.route('/load-data')
def load_data_vulnerable():
    """
    VULNERABLE: Unsafe deserialization of user-provided data
    CWE-502: Deserialization of Untrusted Data
    """
    pickled_data = request.args.get('data')
    data = pickle.loads(pickled_data)  # VULNERABLE
    return str(data)


# ============================================================================
# VULNERABILITY 6: Path Traversal
# ============================================================================

def read_file_vulnerable(filename):
    """
    VULNERABLE: Path traversal via unsanitized filename
    CWE-22: Improper Limitation of a Pathname to a Restricted Directory
    """
    path = f"/var/app/files/{filename}"  # VULNERABLE
    with open(path, 'r') as f:
        return f.read()


def read_file_safe(filename):
    """
    SAFE: Validate filename and restrict to safe directory
    """
    # Ensure filename contains only safe characters
    if not filename.replace('_', '').replace('.', '').isalnum():
        raise ValueError("Invalid filename")

    # Ensure no path separators
    if '/' in filename or '\\' in filename:
        raise ValueError("Path separators not allowed")

    path = os.path.join("/var/app/files", filename)

    # Ensure result is within allowed directory
    if not os.path.abspath(path).startswith("/var/app/files/"):
        raise ValueError("Access denied")

    with open(path, 'r') as f:
        return f.read()


# ============================================================================
# VULNERABILITY 7: Weak Cryptography
# ============================================================================

import hashlib

def hash_password_vulnerable(password):
    """
    VULNERABLE: Using MD5 for password hashing (weak and broken)
    CWE-327: Use of a Broken or Risky Cryptographic Algorithm
    """
    return hashlib.md5(password.encode()).hexdigest()  # VULNERABLE


def hash_password_safe(password):
    """
    SAFE: Using strong password hashing with bcrypt or Argon2
    """
    import bcrypt
    salt = bcrypt.gensalt()
    return bcrypt.hashpw(password.encode(), salt)


# ============================================================================
# VULNERABILITY 8: Information Disclosure via Error Messages
# ============================================================================

@app.route('/user/<user_id>')
def get_user_info_vulnerable(user_id):
    """
    VULNERABLE: Exposing internal details in error messages
    """
    try:
        user = db.get_user(user_id)
        return user.to_json()
    except Exception as e:
        # VULNERABLE: Returning raw error details to user
        return f"Error: {str(e)}", 500  # VULNERABLE


@app.route('/user-safe/<user_id>')
def get_user_info_safe(user_id):
    """
    SAFE: Generic error message, log details internally
    """
    try:
        user = db.get_user(user_id)
        return user.to_json()
    except Exception as e:
        # Log detailed error internally
        app.logger.error(f"Error fetching user {user_id}: {str(e)}")
        # Return generic message to user
        return "An error occurred. Please contact support.", 500


# ============================================================================
# VULNERABILITY 9: Insecure Random Number Generation
# ============================================================================

import random

def generate_token_vulnerable():
    """
    VULNERABLE: Using random module for security-sensitive tokens
    CWE-338: Use of Cryptographically Weak PRNG
    """
    return ''.join([random.choice('abcdef0123456789') for _ in range(32)])  # VULNERABLE


def generate_token_safe():
    """
    SAFE: Using secrets module for cryptographically secure random
    """
    import secrets
    return secrets.token_hex(32)


# ============================================================================
# VULNERABILITY 10: Insecure Direct Object Reference (IDOR)
# ============================================================================

@app.route('/orders/<order_id>')
def get_order_vulnerable(order_id):
    """
    VULNERABLE: No authorization check - any user can access any order
    CWE-639: Insecure Direct Object Reference
    """
    order = db.get_order(order_id)  # VULNERABLE - no authorization check
    return order.to_json()


@app.route('/orders-safe/<order_id>')
def get_order_safe(order_id):
    """
    SAFE: Check that user owns the order before returning it
    """
    order = db.get_order(order_id)
    if order.user_id != current_user.id:
        return "Access denied", 403
    return order.to_json()


if __name__ == '__main__':
    print("This file contains intentional vulnerabilities for testing SAST scanning")
    print("Do NOT use in production!")
